assignment/tool/src/java/org/sakaiproject/assignment/tool/AssignmentAction.java

```diff
@@ -8982,7 +8982,46 @@ private void commitAssignment(SessionState state,
                 a.getGroups().clear();
             } else if (Assignment.Access.GROUP.toString().equals(range)) {
                 a.setTypeOfAccess(Assignment.Access.GROUP);
-                a.setGroups(groups.stream().map(Group::getReference).collect(Collectors.toSet()));
+                Set<String> previousGroupRefs = new HashSet<>(a.getGroups());
+                Set<String> currentGroupRefs = groups.stream().map(Group::getReference).collect(Collectors.toSet());
+
+                // if groups have changed
+                if (!previousGroupRefs.equals(currentGroupRefs)) {
+                    previousGroupRefs.removeAll(currentGroupRefs);
+                    Set<String> submitterIds;
+
+                    if (a.getIsGroup()) {
+                        submitterIds = previousGroupRefs.stream().map(ref -> {
```

Comment:
could you not put all this code inside a.getIsGroup? If it's not a group assignment, does any of this need to be done?

Reply:
generally prefer not to add logic to the database POJO's as its an antipattern, there are better places!

I agree this isn't the ideal place as this logic should live in the service, the only reason i put it here is because this is where similar logic resides. The correct thing would be to move all of this logic into the service but that is a larger effort that this bug.